### ISSUE TITLE
Merge republish strategy leaks files on disk

### ISSUE THREAD

**Issue by mtneug** [opencast/opencast#6035]
**Describe the bug**

When republishing using the merge strategy, distributed files can be "forgotten", effectively leading to a leak on disk / in S3. The merge republish logic is currently something like this:

```java
distribute(updatedMediaPackage.elements);

mergedMediaPackage = updatedMediaPackage;
for (ele in existingMediaPackage.elements) {
  if (ele.flavor in addForceFlavors) {
    mergedMediaPackage.add(ele);
    continue;
  }

  if (ele.flavor in updatedMediaPackage.flavors)
    // replaces existing element with update element
    continue;
  else
    if (ele.flavor in mergeForceFlavors)
      // don't add existing element since this flavor is not in the updated one and we use force
      continue;
    mergedMediaPackage.add(ele);
}

publishToSearch(mergedMediaPackage);

mediaPackage.addEngagePublication(mergedMediaPackage);
```

Note that there is no `retract` for individual media package elements that are not copied over from `existingMediaPackage` to `mergedMediaPackage`. Also, I don't think the documentation is clear on what `merge-force-flavors` and `add-force-flavors` actually do (at least to me, it reads like this deals with elements in `updatedMediaPackage` and not `existingMediaPackage`).

**To Reproduce**

1. Upload video using `schedule-and-upload`
2. After the workflow, update the `partial-publish.xml` workflow definition and change the `publish-engage` operation to the following:
    ```xml
    <operation
      id="publish-engage"
      max-attempts="2"
      exception-handler-workflow="partial-error"
      description="Publishing to Opencast Media Module">
       <configurations>
        <configuration key="download-source-flavors">dublincore/*,security/*</configuration>
        <configuration key="download-source-tags">engage-download,atom,rss</configuration>
        <configuration key="streaming-source-tags">engage-streaming</configuration>
        <configuration key="merge-force-flavors"></configuration>
        <configuration key="strategy">merge</configuration>
        <configuration key="check-availability">false</configuration>
      </configurations>
    </operation>
    ```
3. (Cut and) publish the video using `publish`

After 1. the media package could contain the following elements (depending on the input video resolution, here 720p):
```
Track       d509863d-497c-4901-9bf5-0f806bec8825  presenter/delivery
Track       2230ec64-f92d-4e15-8a6e-78f3bae04769  presenter/delivery
Catalog     309dd5ec-450e-4a13-840b-7a900c15999b  dublincore/episode
Attachment  security-policy-episode               security/xacml+episode
Attachment  444a8fe2-8043-4f52-b7f8-c4d747fc5755  presenter/player+preview
Attachment  73436b70-4c36-48f4-b4c8-3671afa93202  presenter/search+preview
Attachment  5347962d-8f50-4739-b18b-fdd735568ed0  presenter/feed+preview
Attachment  41b87b2a-449d-4b53-babf-89ca854e1a03  presenter/timeline+preview
```

On disk, we have the corresponding files in the download folder:
```
./d509863d-497c-4901-9bf5-0f806bec8825/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f.mp4
./5347962d-8f50-4739-b18b-fdd735568ed0/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-feed.jpg
./309dd5ec-450e-4a13-840b-7a900c15999b/dublincore.xml
./2230ec64-f92d-4e15-8a6e-78f3bae04769/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f.mp4
./73436b70-4c36-48f4-b4c8-3671afa93202/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-search.jpg
./444a8fe2-8043-4f52-b7f8-c4d747fc5755/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-player.jpg
./41b87b2a-449d-4b53-babf-89ca854e1a03/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_0db8e7ce-2a49-49b5-a40a-bc6b0c70e0d7_timelinepreviews.png
./security-policy-episode/xacml.xml
```

After 3. the merged media package replaces every element in the publication:
```
Track       3fb6bfa3-b88f-4baf-a440-f7819c45e996  presenter/delivery
Track       cd7396ab-28d0-4029-bc43-f39c70c7605a  presenter/delivery
Catalog     309dd5ec-450e-4a13-840b-7a900c15999b  dublincore/episode
Attachment  security-policy-episode               security/xacml+episode
Attachment  4e3d12ad-55a3-4455-b716-67e674acde80  presenter/player+preview
Attachment  5ed514d5-a647-4dfb-b2ef-12951b2062ba  presenter/feed+preview
Attachment  fdfa922f-7486-44bc-9ba7-2c24b3a29557  presenter/timeline+preview
Attachment  b217a42c-baea-435b-b96e-8cbd22d1fc98  presenter/search+preview
```

However, old files are still on disk:
```
./4e3d12ad-55a3-4455-b716-67e674acde80/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a_3.000s-player.jpg
./d509863d-497c-4901-9bf5-0f806bec8825/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f.mp4
./5347962d-8f50-4739-b18b-fdd735568ed0/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-feed.jpg
./b217a42c-baea-435b-b96e-8cbd22d1fc98/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a_3.000s-search.jpg
./309dd5ec-450e-4a13-840b-7a900c15999b/dublincore.xml
./5ed514d5-a647-4dfb-b2ef-12951b2062ba/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a_3.000s-feed.jpg
./2230ec64-f92d-4e15-8a6e-78f3bae04769/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f.mp4
./fdfa922f-7486-44bc-9ba7-2c24b3a29557/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a_faa7f9b5-050e-414a-ab81-94148405ebf1_timelinepreviews.png
./3fb6bfa3-b88f-4baf-a440-f7819c45e996/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a.mp4
./73436b70-4c36-48f4-b4c8-3671afa93202/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-search.jpg
./444a8fe2-8043-4f52-b7f8-c4d747fc5755/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_3.000s-player.jpg
./41b87b2a-449d-4b53-babf-89ca854e1a03/07d6b31d-0bd0-4f6f-906c-4fa4b342ae5f_0db8e7ce-2a49-49b5-a40a-bc6b0c70e0d7_timelinepreviews.png
./cd7396ab-28d0-4029-bc43-f39c70c7605a/presenter-7c5ba9c0-a0ea-4d87-bdc5-fd3caa56ba1a.mp4
./security-policy-episode/xacml.xml
```

Note that the ACLs and dublincore elements are not duplicated since the element IDs stay the same and the distribution service simply overwrites them. However, this behavior is never guaranteed by the distribution service and it might change. As such, the `merge` strategy can currently only be used if the flavor of all new elements is not already used in the old publication.

**Expected behavior**

Old files are retracted.

**Server environment:**
 - OS: Container
 - Distribution: `allinone`
- Installation method: container
 - Version: 16.1

**Comment by KatrinIhler** [comment]
Imo this is a duplicate of https://github.com/opencast/opencast/issues/5963

**Comment by mtneug** [comment]
Closed in favor of #5963.
